packages/flutter_tools/lib/src/asset.dart

```diff
@@ -487,42 +494,46 @@ Map<_Asset, List<_Asset>> _parseAssets(
   return result;
 }
 
-_Asset _resolvePackageAsset(
-    String assetBase,
-    String packageName,
-    String asset,
-) {
-  return new _Asset(
-    base: assetBase,
-    assetEntry: 'packages/$packageName/$asset',
-    relativePath: asset,
-  );
-}
-
 _Asset _resolveAsset(
   PackageMap packageMap,
   String assetBase,
   String asset,
+  String packageName,
 ) {
   if (asset.startsWith('packages/') && !fs.isFileSync(fs.path.join(assetBase, asset))) {
-    // Convert packages/flutter_gallery_assets/clouds-0.png to clouds-0.png.
-    String packageKey = asset.substring(9);
-    String relativeAsset = asset;
-
-    final int index = packageKey.indexOf('/');
-    if (index != -1) {
-      relativeAsset = packageKey.substring(index + 1);
-      packageKey = packageKey.substring(0, index);
-    }
+    // The asset is referenced in the pubspec.yaml as
+    // 'packages/PACKAGE_NAME/PATH/TO/ASSET
+    final _Asset packageAsset = _resolvePackageAsset(asset, packageMap);
+    if (packageAsset != null)
+      return packageAsset;
+  }
 
-    final Uri uri = packageMap.map[packageKey];
-    if (uri != null && uri.scheme == 'file') {
-      final File file = fs.file(uri);
-      return new _Asset(base: file.path, assetEntry: asset, relativePath: relativeAsset);
-    }
+  final String assetEntry = packageName != null
+    ? 'packages/$packageName/$asset' // Asset from, and declared in $packageName
+    : null; // Asset from the current application
+  return new _Asset(base: assetBase, assetEntry: assetEntry, relativePath: asset);
+}
+
+_Asset _resolvePackageAsset(String asset, PackageMap packageMap) {
+  assert(asset.startsWith('packages/'));
+  String packageKey = asset.substring(9);
+  String relativeAsset = asset;
+
+  final int index = packageKey.indexOf('/');
+  if (index != -1) {
```

Comment:
If `asset` is `packages/vacation/sunset.jpg`, we get
```
relativeAsset == 'sunset.jpg'
packageKey == 'vacation'
```
That seems reasonable. But if `asset` is `packages/vacation_sunset.jpg`, we get
```
relativeAsset == 'packages/vacation_sunset.jpg'
packageKey == 'vacation_sunset.jpg'
```
That appears backwards to me?